### DISCUSSION TITLE
Allow free-form queries within FastAPI

### DISCUSSION THREAD

**Discussion by yeralin** [fastapi/fastapi#9075] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description
There is a notion of "free-form query parameters" that allows a user to accept multiple non-predefined query params:
![image](https://user-images.githubusercontent.com/8620461/84204081-9329a980-aa78-11ea-9618-18d4f9dd5387.png)

This is in fact can be described by Open-API 3.0:
https://stackoverflow.com/questions/49582559/how-to-document-dynamic-query-parameter-names-in-openapi-swagger

Unfortunately, as of now you cannot define "free-form query parameters" within FastAPI.

### Example

This is how it would look like:

```Python
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def read_root(arguments: Dict[str, str] = Query({})):
    ...
```

But unfortunately, this raises an exception:
```
AssertionError: Param: arguments can only be a request body, using Body(...)
```

While at it, I hacked FastAPI a bit, and managed to make it render docs page with the valid entries:
![image](https://user-images.githubusercontent.com/8620461/84204398-1fd46780-aa79-11ea-9e27-1cffc0f86206.png)
By modifying: `sequence_shapes` set:
```
from fastapi.dependencies.utils import sequence_shapes
from pydantic.fields import SHAPE_MAPPING

sequence_shapes.add(SHAPE_MAPPING)
```
I understand that it is not how it should be done, I just wanted to verify that it would generate a valid `openapi.json`. 

Also not surprisingly, `arguments` var would evaluate to an empty dict `{}`.

### Describe alternatives you've considered

Currently, I am using the following workaround:
```Python
from pydantic.types import Json
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def read_root(arguments: Json = Query({})):
    ...
```

![image](https://user-images.githubusercontent.com/8620461/84204844-ef40fd80-aa79-11ea-8598-b2f1583ecb7e.png)

That allows me to pass multiple query parameters that are only assigned to `arguments` `Json|Dict` object. However, all params needs to be encoded inside `?arguments=<params>` which is undesirable.

### Environment

* OS: macOS:Darwin Kernel Version 19.5.0
* FastAPI Version: 0.55.1

* Python version: 3.7.7

### Additional context

I am willing to prepare a PR if provided with some pointers :)

**Comment by YKo20010** [comment]
@chrisngyn and @YKo20010 will be working on this!

**Comment by yeralin** [comment]
Awesome! I can contribute too, if you need any help :)

**Comment by YKo20010** [comment]
@yeralin 
Hey Daniyar, @chrisngyn and I are wrapping up some of our other open issues before our fellowship ends, so feel free to take this one on if you'd like!
This is the contributions page to help you get started: https://fastapi.tiangolo.com/contributing/ 
And here's what our maintainer @tiangolo had in mind, so it may take a bit of experimenting: 
<img width="677" alt="image" src="https://user-images.githubusercontent.com/43295248/89920582-92bda400-dbca-11ea-8372-820125895eda.png">

**Comment by yeralin** [comment]
Created a PR with an initial idea on how this feature would look like:
https://github.com/tiangolo/fastapi/pull/1885

@YKo20010 @chrisngyn @tiangolo

**Comment by mikeedjones** [comment]
I picked up @yeralin 's PR and merged the changes added since 2020. My PR is #9255 @YKo20010 @tiangolo

**Reply by mikeedjones** [reply]
Have updated the PR to support pydantic v2 and fastAPI v 0.100 https://github.com/tiangolo/fastapi/pull/9837 - made a new PR and closed the old

**Comment by mikeedjones** [comment]
@tiangolo this feature request has been open for a while, and pr addressing it was recently closed. Is there any plan to implement this feature of swagger in the upcoming reactors?

Cheers for all the work!

**Comment by boubakerdouz96-design** [comment]
- [x] !- [ ] - [ ] - [ ] - [ ] [FB_IMG_1777672741090.jpg](https://github.com/user-attachments/assets/45439909-bcbd-445d-8849-df112bc851d6)
